server/src/test/java/org/apache/druid/segment/indexing/DataSchemaTest.java

```diff
@@ -324,6 +324,43 @@ public void testEmptyDatasource()
     );
   }
 
+  @Test
+  public void testNewlineDatasource()
+  {
+    Map<String, Object> parser = jsonMapper.convertValue(
+        new StringInputRowParser(
+            new JSONParseSpec(
+                new TimestampSpec("time", "auto", null),
+                new DimensionsSpec(
+                    DimensionsSpec.getDefaultSchemas(ImmutableList.of("time", "dimA", "dimB", "col2")),
+                    ImmutableList.of("dimC"),
+                    null
+                ),
+                null,
+                null
+            ),
+            null
+        ), JacksonUtils.TYPE_REFERENCE_MAP_STRING_OBJECT
+    );
+
+    expectedException.expect(CoreMatchers.instanceOf(IllegalArgumentException.class));
+    expectedException.expectMessage(
+        "dataSource cannot contain the '/n' character."
```

Comment:
Please add a similar test case for each of the other new forbidden characters (\t and \r).